src/coord/src/sink_connector.rs

```diff
@@ -107,49 +107,41 @@ fn get_latest_ts(
         )
     })?;
 
-    let mut latest_message = None;
-    while let Some(message) = get_next_message(consumer, timeout)? {
-        latest_message = Some(message);
+    let (hi, lo) = consumer
+        .fetch_watermarks(consistency_topic, 0, timeout)
+        .map_err(|e| {
+            anyhow!(
+                "Failed to fetch metadata while reading from consistency topic: {}",
+                e
+            )
+        })?;
+    let hi: u64 = hi.try_into()?;
+
+    // Empty topic
+    if hi == 0 {
+        return Ok(None);
     }
 
-    if latest_message.is_none() {
-        // fetch watermarks to distinguish between a timeout reading end-1 and an empty topic
-        match consumer.fetch_watermarks(consistency_topic, 0, timeout) {
-            Ok((lo, hi)) => {
-                if hi == 0 {
-                    return Ok(None);
-                } else {
-                    bail!(
-                        "uninitialized consistency topic {}:{}, lo/hi: {}/{}",
-                        consistency_topic,
-                        partition,
-                        lo,
-                        hi
-                    );
-                }
-            }
-            Err(e) => {
-                bail!(
-                    "Failed to fetch metadata while reading from consistency topic: {}",
-                    e
-                );
-            }
+    while let Some(message) = get_next_message(consumer, timeout)? {
+        if Some(hi) == maybe_decode_consistency_end_record(&message, consistency_topic)? {
```

Comment:
This check won't work: `hi` is the offset of the latest message in the consistency topic while the timestamp returned from `maybe_decode_consistency_end_record()` is the latest timestamp that we have completely written out to the data topic.

What you would have to do is thread the message offset from `get_next_message()` (which you can get via `message.offset()`) back to here and compare against that offset. And then you would keep updating `latest_message` with the currently read message until we arrive at the last one.

One complication is that the high watermark might not necessarily point to the last record in the consistency topic. It can happen that there is "garbage" at the end of a topic. I don't know if `hi` would point at the garbage in that case, or the latest valid record.

Reply:
I didn't really understand the suggestion here @aljoscha is this diff roughly what you were thinking of

```
diff --git a/src/coord/src/sink_connector.rs b/src/coord/src/sink_connector.rs
index da7412c79..6a6f68f07 100644
--- a/src/coord/src/sink_connector.rs
+++ b/src/coord/src/sink_connector.rs
@@ -46,17 +46,20 @@ pub async fn build(
 fn get_next_message(
     consumer: &mut BaseConsumer,
     timeout: Duration,
-) -> Result<Option<Vec<u8>>, anyhow::Error> {
+) -> Result<Option<(Vec<u8>, i64)>, anyhow::Error> {
     if let Some(result) = consumer.poll(timeout) {
         match result {
-            Ok(message) => match message.payload() {
-                Some(p) => Ok(Some(p.to_vec())),
-                None => {
-                    bail!("unexpected null payload")
+            Ok(message) => {
+                let offset = message.offset();
+                match message.payload() {
+                    Some(p) => Ok(Some((p.to_vec(), offset))),
+                    None => {
+                        bail!("unexpected null payload");
+                    }
                 }
-            },
+            }
             Err(err) => {
-                bail!("Failed to process message {}", err)
+                bail!("Failed to process message {}", err);
             }
         }
     } else {
@@ -107,7 +110,7 @@ fn get_latest_ts(
         )
     })?;

-    let (hi, lo) = consumer
+    let (hi, _) = consumer
         .fetch_watermarks(consistency_topic, 0, timeout)
         .map_err(|e| {
             anyhow!(
@@ -115,26 +118,43 @@ fn get_latest_ts(
                 e
             )
         })?;
-    let hi: u64 = hi.try_into()?;
+    let hi: i64 = hi.try_into()?;

     // Empty topic
     if hi == 0 {
         return Ok(None);
     }

-    while let Some(message) = get_next_message(consumer, timeout)? {
-        if Some(hi) == maybe_decode_consistency_end_record(&message, consistency_topic)? {
-            return Ok(Some(hi));
+    let mut latest_ts = None;
+    let mut latest_offset = 0;
+    while let Some((message, offset)) = get_next_message(consumer, timeout)? {
+        debug_assert!(offset >= latest_offset);
+        latest_offset = offset;
+        // We don't return an error if we cannot decode the END records because there
+        // may be arbitrary messages in this topic that we cannot decode.
+        if let Ok(res) = maybe_decode_consistency_end_record(&message, consistency_topic) {
+            match (res, &mut latest_ts) {
+                (Some(ts), Some(curr_latest)) => {
+                    debug_assert!(ts > *curr_latest);
+                    *curr_latest = ts;
+                }
+                (Some(ts), None) => latest_ts = Some(ts),
+                _ => continue,
+            }
         }
     }

-    bail!(
-        "uninitialized consistency topic {}:{}, lo/hi: {}/{}",
-        consistency_topic,
-        partition,
-        lo,
-        hi
-    );
+    if latest_offset != hi {
+        // TODO: unclear to me if this is really an error.
+        bail!(
+            "could not re-read all of consistency topic {}: read to {} expected {}",
+            consistency_topic,
+            latest_offset,
+            hi
+        );
+    }
+
+    Ok(latest_ts)
 }

 // Errors if cannot decode message; returns `None` if not `END` message.
```

basically, read through the topic, keeping track of the largest decoded end ts you see along the way, and error out if you couldn't actually read the whole topic?

Reply:
and don't error if you can't decode END records?

Reply:
I guess my thought was you want to be sure you've at least scanned all the offsets, not because the end record is necessarily at the last offset, but because we need to be sure we haven't exited early and missed an end record.

Reply:
With this diff (and aljoscha's suggestion), the only situation where we avoid the 5 second timeout is when reading from an empty sink (`if hi == 0 { return Ok(None)}`) which seems like a good improvement but I think we can do better.

It sounds like we should be able to break out of the loop early if we _happen_ to read a message with offset equalling the `hi` watermark -- without requiring this to the state of the kafka topic.

I'll put up another iteration now that you've pointed out some flawed assumptions I've had.